it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,219 @@
+import com.google.cloud.spanner.DatabaseId;
+import com.google.cloud.spanner.InstanceAdminClient;
+import com.google.cloud.spanner.InstanceConfigId;
+import com.google.cloud.spanner.InstanceId;
+import com.google.cloud.spanner.InstanceInfo;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+import com.google.cloud.teleport.it.spanner.SpannerResourceManagerException.CreateDatabaseException;
+import com.google.cloud.teleport.it.spanner.SpannerResourceManagerException.CreateInstanceException;
+import com.google.cloud.teleport.it.spanner.SpannerResourceManagerException.UpdateDatabaseException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.CharMatcher;
+import com.google.common.collect.ImmutableList;
+import com.google.common.hash.HashFunction;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.ExecutionException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default class for implementation of {@link SpannerResourceManager} interface.
+ *
+ * <p>There are one instance, one database, and multiple tables per manager object. The tables are
+ * under the database, and the database is under the instance. Instance creation is handled by
+ * {@link Builder#build()} using {@link DefaultSpannerResourceManager#createInstance(String,
+ * String)}. Database creation is triggered by the {@link
+ * DefaultSpannerResourceManager#createTable(String)} method, if there is no existing database.
+ *
+ * <p>The instance and database ids are formed using testId. The database id string will be
+ * {testId}, with some extra formatting. The instance id will be "{testId}-{ISO8601 time,
+ * microsecond precision}", with additional formatting.
+ *
+ * <p>Note: If testId is more than 30 characters, a new testId will be formed for naming: {first 22
+ * chars of long testId} + “-” + {7 char hash of testId}.
+ */
+public class DefaultSpannerResourceManager implements SpannerResourceManager {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSpannerResourceManager.class);
+
+  private final InstanceId instanceId;
+  private final DatabaseId databaseId;
+
+  private final Spanner spanner;
+  private final InstanceAdminClient instanceAdminClient;
+  private final DatabaseAdminClient databaseAdminClient;
+  private DatabaseClient databaseClient;
+
+  private DefaultSpannerResourceManager(Builder builder) {
+    this(builder.projectId, builder.instanceId, builder.databaseId);
+  }
+
+  @VisibleForTesting
+  DefaultSpannerResourceManager(String projectId, String instanceId, String databaseId) {
+    this.instanceId = InstanceId.of(projectId, instanceId);
+    this.databaseId = DatabaseId.of(projectId, instanceId, databaseId);
```

Comment:
Having this here could cause the constructor to fail. If there is an invalid instance id or database id, InstanceId.of() and DatabaseId.of() will fail. The alternative is to make projectId a class property, that way, I can use DatabaseId.of() in createTable(). Should I or do I leave it as is?